InformationScripting/src/graph/Graph.cpp

```diff
@@ -0,0 +1,191 @@
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "Graph.h"
+
+#include "InformationNode.h"
+#include "InformationEdge.h"
+
+namespace InformationScripting {
+
+QList<Graph::NodeHash> Graph::nodeHashFunctions_;
+
+Graph::~Graph()
+{
+	for (auto& e : edges_) SAFE_DELETE(e);
+	for (auto& n : nodes_) SAFE_DELETE(n);
+}
+
+InformationNode* Graph::add(InformationNode* node)
+{
+	Q_ASSERT(!nodeHashFunctions_.empty());
+
+	InformationNode* existingNode = findNode(node);
+
+	if (existingNode && existingNode != node)
+	{
+		mergeNodes(existingNode, node);
+		SAFE_DELETE(node);
+	}
+	else
+	{
+		nodes_[hashValueOf(node)] = node;
+		existingNode = node;
+	}
+	return existingNode;
+}
+
+InformationEdge* Graph::addDirectedEdge(InformationNode* from, InformationNode* to, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(from == findNode(from));
+	Q_ASSERT(to == findNode(to));
+	for (auto edge : edges_)
+	{
+		if (edge->from() == from && edge->to() == to && edge->name() == name)
+		{
+			auto existingEdge = edge;
+			Q_ASSERT(existingEdge->isDirected());
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(from, to, name);
+	edges_.push_back(edge);
+	return edge;
+}
+
+InformationEdge* Graph::addEdge(InformationNode* a, InformationNode* b, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(a == findNode(a));
+	Q_ASSERT(b == findNode(b));
+	for (auto edge : edges_)
+	{
+		if ((edge->from() == a || edge->from() == b) && (edge->to() == b || edge->to() == a) && edge->name() == name)
+		{
+			auto existingEdge = edge;
+			Q_ASSERT(!existingEdge->isDirected());
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(a, b, name, InformationEdge::Orientation::Undirected);
+	edges_.push_back(edge);
+	return edge;
+}
+
+void Graph::remove(InformationNode* node)
+{
+	Q_ASSERT(node);
+	auto hash = hashValueOf(node);
+	auto nodeIt = nodes_.find(hash);
+	// If the node isn't in the graph we don't have to do anything:
+	if (nodeIt == nodes_.end()) return;
+
+	// It could be that node points to a different node with the same hash.
```

Comment:
We should discuss the whole hashing/equality approach in more detail. A problem is that we can add properties to nodes, but that should not affect the hash value, otherwise we lose the uniqueness property, e.g. A node which contains only the AST property with method X should be seen as the same as A node which contains the AST property with method X and additional properties.